### PR TITLE
Closes #34 - Add native Go SDK

### LINKED ISSUE #34: Golang SDK immudb
Actually immugw rest proxy exposes simplify api that wrap immudb one's.
All safe methods exposes by immugw (SafeSet, SafeGet, SafeReference, SafeZAdd) doesn't require for root and index because this data is managed internally.
The purpose of this task is to create a golang SDK to offer the same capabilities but in pure GO.
It should be fine that immugw wil be upgraded in order to use the newly sdk.
In order to do this you could use a new .proto schema that extends https://github.com/codenotary/immudb/blob/master/pkg/api/schema/schema.proto 
Ex:
actual SafeGet service in immudb schema.proto
```
message SafeGetOptions {
	Key key = 1;
	Index rootIndex = 2;
}
rpc SafeGet(SafeGetOptions) returns (SafeItem){
		option (google.api.http) = {
            post: "/v1/immurestproxy/item/safe/get"
            body: "*"
        };
	};
```
New SafeSet in immu.proto
```
rpc SafeGet(Key) returns (VeridiedItem){};
```
eg: VerifiedItem contains item and a bool in which we put verification results.

Flow:
- The sdk will intercept SafeGet,
- will retrieve rootIndex from client.RootService package
- will forward the call to immudb (with SafeGet(SafeGetOptions) )
- will make checks on result
- will return to the user the VerifiedItem

### PR DESCRIPTION
**Pull request by padurean**
Also add safeget, safeset, safereference, safezadd, backup and restore commands to CLI

### CODE REVIEW

File: pkg/client/client.go [hunk 1]
```diff
@@ -267,12 +447,245 @@ func (c *ImmuClient) ZAdd(setReader io.Reader, score float64, keyReader io.Reade
...
+
+	bkpClient, err := c.serviceClient.Backup(ctx, &empty.Empty{})
+	if err != nil {
+		return counter, err
+	}
+
+	var offset int64
+	var errs []string
+	for {
+		kvList, err := bkpClient.Recv()
```
Comment (mmeloni) [inline pkg/client/client.go:570]: You need to close stream.
Take a look at https://grpc.io/docs/tutorials/basic/go/
stream.SendAndClose

File: pkg/client/client_test.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
+package client
```
Comment (mmeloni) [inline pkg/client/client_test.go:1]: add License

File: pkg/client/client.go [hunk 3]
```diff
@@ -167,98 +310,135 @@ func (c *ImmuClient) SetBatch(request *BatchRequest) (*schema.Index, error) {
...
 		Key:       &schema.Key{Key: key},
 	})
 	c.Logger.Debugf("reference finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) ZAdd(setReader io.Reader, score float64, keyReader io.Reader) (*schema.Index, error) {
+// SafeReference ...
+func (c *ImmuClient) SafeReference(ctx context.Context, reference []byte, key []byte) (*VerifiedIndex, error) {
 	start := time.Now()
```
Comment (mmeloni) [inline pkg/client/client.go:381]: All safe methods need to be thread safe.
Use mutex to lock immuclient properly, and release lock when methods return

**Comment by padurean** [comment]
@mmeloni I've implemented all the changes/fixes requested. Thanks and please let me know if you notice anything else.

**Comment by padurean** [comment]
@mmeloni beside the failing check on linux (which doesn’t seem to be related to code), there’s one merge commit that needs to be cleared from the log before merging it into `master` branch.

**Pull request merged**